File: C_src/medium/238_Product_of_Array_Except_Self/ProductOfArrayExceptSelf_238.c
#include <stdio.h>
#include <stdlib.h>

// Use 1 parameter and result array record:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 56 ms, faster than 28.28%
// Memory Usage: 15.3 MB, less than 100.00%
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *productExceptSelf(int *nums, int numsSize, int *returnSize)
{
    int *result = calloc(sizeof(int), *returnSize = numsSize);

    *result = 1;
    for (int i = 1; i < numsSize; i++)
    {
        *(result + i) = *(result + i - 1) * *(nums + i - 1);
    }

    int R = 1;
    for (int i = numsSize - 1; i >= 0; i--)
    {
        *(result + i) *= R;
        R *= *(nums + i);
    }

    return result;
}

int *productExceptSelf2(int *nums, int numsSize, int *returnSize)
{
    int *result = calloc(sizeof(int), *returnSize = numsSize);

    result[0] = 1;
    for (int i = 1; i < numsSize; i++)
    {
        result[i] = result[i - 1] * nums[i - 1];
    }

    int R = 1;
    for (int i = numsSize - 1; i >= 0; i--)
    {
        result[i] *= R;
        R *= nums[i];
    }

    return result;
}

File: Golang_src/medium/productOfArrayExceptSelf_238.go
package medium

// Use Left and Right product lists:
// Time Complexity: O(3n) = O(n)
// Space Complexity:O(2n) = O(n)
// Runtime: 28 ms, faster than 29.00%
// Memory Usage: 6.6 MB, less than 37.13%
func productExceptSelf(nums []int) []int {
	size := len(nums)
	left := make([]int, size)
	right := make([]int, size)
	left[0] = 1
	for i := 1; i < size; i++ {
		left[i] = nums[i-1] * left[i-1]
	}
	right[size-1] = 1
	for i := size - 2; i >= 0; i-- {
		right[i] = nums[i+1] * right[i+1]
	}

	result := make([]int, size)
	for i := 0; i < size; i++ {
		result[i] = left[i] * right[i]
	}

	return result
}

// Optimize Left and Right product lists:
// Time Complexity: O(2n) = O(n)
// Space Complexity:O(1)
// Runtime: 28 ms, faster than 29.00%
// Memory Usage: 6.9 MB, less than 37.13%
func productExceptSelf2(nums []int) []int {
	size := len(nums)
	left := make([]int, size)

	left[0] = 1
	for i := 1; i < size; i++ {
		left[i] = nums[i-1] * left[i-1]
	}

	tmp := 1
	for i := size - 2; i >= 0; i-- {
		tmp *= nums[i+1]
		left[i] = tmp * left[i]
	}

	return left
}
